lightningd/peer_htlcs.c

```diff
@@ -1842,3 +1842,80 @@ static const struct json_command dev_ignore_htlcs = {
 };
 AUTODATA(json_command, &dev_ignore_htlcs);
 #endif /* DEVELOPER */
+
+static void listforwardings_add_stats(struct json_result *response, struct wallet *wallet)
+{
+	const struct forwarding_stats *stats;
+	stats = wallet_forwarded_payments_stats(wallet, tmpctx);
+
+	json_object_start(response, "stats");
+	json_add_num(response, "settled_fee_msatoshis", stats->fee[FORWARD_SETTLED]);
+	json_add_num(response, "settled_count", stats->count[FORWARD_SETTLED]);
+	json_add_num(response, "settled_msatoshi", stats->msatoshi[FORWARD_SETTLED]);
+
+	json_add_num(response, "failed_fee_msatoshis", stats->fee[FORWARD_FAILED]);
+	json_add_num(response, "failed_count", stats->count[FORWARD_FAILED]);
+	json_add_num(response, "failed_msatoshi", stats->msatoshi[FORWARD_FAILED]);
+
+	json_add_num(response, "pending_fee_msatoshis", stats->fee[FORWARD_OFFERED]);
+	json_add_num(response, "pending_count", stats->count[FORWARD_FAILED]);
+	json_add_num(response, "pending_msatoshi", stats->msatoshi[FORWARD_FAILED]);
+	json_object_end(response);
+
+	tal_free(stats);
+}
+
+static void listforwardings_add_forwardings(struct json_result *response, struct wallet *wallet)
+{
+	const struct forwarding *forwardings;
+	forwardings = wallet_forwarded_payments_get(wallet, tmpctx);
+
+	json_array_start(response, "forwardings");
+	for (size_t i=0; i<tal_count(forwardings); i++) {
+		const struct forwarding *cur = &forwardings[i];
+		json_object_start(response, NULL);
+		json_add_short_channel_id(response, "in_channel", &cur->channel_in);
+		json_add_short_channel_id(response, "out_channel", &cur->channel_out);
+		json_add_num(response, "in_msatoshi", cur->msatoshi_in);
+		json_add_num(response, "out_msatoshi", cur->msatoshi_out);
+		json_add_num(response, "fee", cur->fee);
+		switch (cur->status) {
+		case FORWARD_OFFERED:
+			json_add_string(response, "status", "offered");
+			break;
+		case FORWARD_SETTLED:
+			json_add_string(response, "status", "settled");
+			break;
+		case FORWARD_FAILED:
+			json_add_string(response, "status", "failed");
+			break;
+		}
+		json_object_end(response);
+	}
+	json_array_end(response);
+
+	tal_free(forwardings);
+}
+
+static void json_listforwardings(struct command *cmd, const char *buffer,
+				 const jsmntok_t *params)
+{
+	struct json_result *response = new_json_result(cmd);
+
+	if (!param(cmd, buffer, params, NULL))
+		return;
+
+	json_object_start(response, NULL);
+	listforwardings_add_stats(response, cmd->ld->wallet);
+	listforwardings_add_forwardings(response, cmd->ld->wallet);
+	json_object_end(response);
```

Comment:
Interesting, my main goal was to have those stats exposed and have the detailed list be options. lnd is calling it fwdhistory btw.

Reply:
I generally think that exposing stats is wrong, since people want percentiles and histograms etc and you'll always get something wrong. Plugins are the right place for this :)